enterprise/backend/src/metabase_enterprise/workspaces/driver/postgres.clj

```diff
@@ -0,0 +1,74 @@
+(ns metabase-enterprise.workspaces.driver.postgres
+  "Postgres-specific implementations for workspace isolation."
+  (:require
+   [clojure.java.jdbc :as jdbc]
+   [metabase-enterprise.workspaces.driver.common :as driver.common]
+   [metabase-enterprise.workspaces.isolation :as isolation]
+   [metabase-enterprise.workspaces.sync :as ws.sync]
+   [metabase.driver.sql-jdbc.connection :as sql-jdbc.conn]
+   [metabase.driver.util :as driver.u]))
+
+(defmethod isolation/grant-read-access-to-tables! :postgres
+  [database username tables]
+  (let [driver         (driver.u/database->driver database)
+        jdbc-spec      (sql-jdbc.conn/connection-details->spec driver (:details database))
+        sqls           (->> (for [table tables]
+                              [(format "GRANT USAGE ON SCHEMA %s TO %s" (:schema table) username)
+                               (format "GRANT SELECT ON TABLE %s.%s TO %s" (:schema table) (:name table) username)])
+                            flatten
+                            ;; drop mutliple grant usage on the same schema
+                            distinct)]
```

Comment:
Flatten is my most hated function, almost always a smell. I also don't like generating all these duplicates and then doing so much extra set membership checking.

How about something like this instead:

```clojure
(concat (map grant-schema-statement (distinct (map :schema tables)))
        (map grant-table-statement tables))
```

Still has a distinct check, but doesn't allocate nearly as much, and in the limit uses a massively smaller set for the membership checks.

Reply:
You could also avoid the concat and just doseq each collection separately.

Concat also has some memory foot guns, but can't recall them all (might not matter here)